{"cells":[{"source":"<a href=\"https://www.kaggle.com/code/ayushs9020/re-creating-protein-neural-nets-from-scratch?scriptVersionId=135937784\" target=\"_blank\"><img align=\"left\" alt=\"Kaggle\" title=\"Open in Kaggle\" src=\"https://kaggle.com/static/images/open-in-kaggle.svg\"></a>","metadata":{},"cell_type":"markdown"},{"cell_type":"markdown","id":"aaaacab5","metadata":{"papermill":{"duration":0.014718,"end_time":"2023-07-06T16:04:05.747804","exception":false,"start_time":"2023-07-06T16:04:05.733086","status":"completed"},"tags":[]},"source":["# Protein Neural Network\n","\n","$Protein$ $Neural$ $Network$ is a type of `deep learning neural network` that is used to `predict` the `properties of proteins`. They are `trained` on a `large dataset` of `protein sequences`/`structures`. The network learns to `identify patterns` in the data that are associated with `particular properties`. $Protein$ $Neural$ $Networks$ have been shown to be `very effective` at `predicting protein properties`. In many cases, they `outperform traditional` methods that rely on `hand-crafted features`\n","\n","<img src = \"https://scontent.fjai2-1.fna.fbcdn.net/v/t1.6435-9/37298434_654990401529724_543197132539035648_n.jpg?_nc_cat=102&ccb=1-7&_nc_sid=730e14&_nc_ohc=7hJcasqo1PkAX8fs55i&_nc_ht=scontent.fjai2-1.fna&oh=00_AfBLiYMGoJgSMX0w-kd690Y2GvRX4ga07ER2OXTSeH7o5w&oe=64CE544A\" width = 400>\n","\n","\n","# 1 | Basic Terminologies 💻\n","\n","* $Structure$ $of$ $a$ $Protein$\n","* $Gene$ $Ontology$ $(GO)$\n","* $Principal$ $Component$ $Analaysis$\n","* $Transformers$\n","\n","## 1.1 | Structure of a Protein\n","\n","So what is a actually a **Protein...?**\n","\n","First of all lets understand the structure of an **Atom**\n","\n","<img src = \"https://www.sciencefacts.net/wp-content/uploads/2020/11/Parts-of-an-Atom-Diagram.jpg\"  width = 300>\n","\n","There is a really good image I found of the `structure of atom`. Though there are many debates on the structure like this, but this `model is accepted universaly at this moment`.\n","\n","In the centre we have the `Neucleus`. The `Neucleus` is made up of $2$ more structures named as `Neutron` and `Proton`. A `Proton` is `positively charged element` and a `Neutron` is a `neutral charged element`. A `Electron`, `negatively charged element`, `orbits` this `Neucleus` at some `distance apart`.\n","\n","The `more we increase the number` of `Electrons` and `Protons`. The `bigger the atoms becomes`.\n","\n","There are `different shells` where the `Electrons reside`. The `more closer the shell` is, the `less Electrons` it contrains. There are mainly $4$ shells. \n","\n","|||\n","|---|---\n","|$K$|$2$\n","|$L$|$8$\n","|$M$|$18$\n","|$N$|$32$\n","\n","Once an atom `fills its outer most shell` with `Electrons`. It becomes `stable atom` and try to `refuse any donation` or `recieve of extra atom`.\n","\n","<img src = \"https://cdn1.byjus.com/wp-content/uploads/2022/01/word-image128.png\" width = 400>\n","\n","`Different atoms combine` to `share Electrons` and become `stable Molecules` \n","\n","<img src = \"https://www.astrochem.org/sci_img/Amino_Acid_Structure.jpg\" width = 300>\n","\n","A `Amino Acid` is made up of mainly $4$ different atoms\n","`[H , C , O , N]`. \n","\n","<img src = \"https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/L-amino_acid_structure.svg/1200px-L-amino_acid_structure.svg.png\" alt = \"Bro use Light Theme\" width = 300 >\n","\n","We also have a free `Electron Pair` of `Carbon` in this molecule, we call this as a `Side Chain` which can be of different types. Basically this `Side Chain` provide the flexibility to make `different types` of `Amino Acids`. This flexibilty allows for $20$ `different` `Amino Acids` \n","\n","When we join `Amino Acids` with `peptide bonds`, we get `Proteins`. Conncecting different types of `Amino Acids` ends up in different types of `Proteins`.\n","\n","## 1.2 | Gene Ontology (GO)\n","\n","$Gene$ $Ontology$ $(GO)$ is a `controlled vocabulary` that `describes the functions of genes` and gene products. It is constantly being updated as new information becomes available.\n","\n","There are mainly $3$ `Ontologies`\n","\n","||||\n","|---|---|---\n","|$Biological$ $Process$|Describes the `biological processes`|A gene product might be involved in the process of `cell cycle`/`signal transduction.`\n","|$Cellular$ $Component$|Describes the `cellular components`|A gene product might be located in the `nucleus`/`cytoplasm`.\n","|$Molecular$ $Function$|Describes the `molecular functions`|A gene product might be involved in the `catalysis of a reaction`/`binding of a molecule`.\n","\n","**[Gene Ontology Documentation](http://geneontology.org/docs/ontology-documentation/)**\n","\n","## 1.3 | Principal Component Analysis\n","\n","<img src = \"https://media.makeameme.org/created/pca-the-cause.jpg\" width = 400>\n","\n","$Principal$ $Component$ $Analysis$ $PCA$ is a `statistical procedure` that uses an `orthogonal transformation` to convert a set of `correlated variables` into a set of `uncorrelated variables` called principal components.\n","\n","First we compute the Covariance Matrix with the formula \n","\n","$$Cov_{x , y} = \\frac{\\sum(x_i - x_{mean})(y_i - y_{mean})}{N-1}$$\n","\n","Then we find the Eign Values and Eign Vectors \n","\n","## 1.4 | Transformers \n","\n","A $Transformer$ is a `deep learning architecture` that relies on the `attention mechanism`. It is notable for requiring `less training time`. The model takes in `tokenized` `byte pair encoding` `input` tokens, and at each layer, `contextualizes each token` with other (unmasked) input tokens in `parallel` via attention mechanism.\n","\n","* $Attention$ $Mechanism$- The attention mechanism allows the transformer to `learn long-range dependencies` between `tokens` in a sequence. \n","* $Self-Attention$ - Self-attention is a special type of attention that allows the transformer to `attend to itself`. This means that the transformer can learn relationships between different parts of the same sequence.\n","* $Encoder-Decoder$ $Architecture$ - The transformer is an `Encoder-Decoder` architecture. This means that the transformer has two parts\n","* * Encoder - takes the input sequence and produces a sequence of hidden states\n","* * Decoder - takes these hidden states and produces the output sequence.\n","\n","$$a = Softmax(\\frac{KQ^T}{\\sqrt{d_k}})$$\n","\n","<img src = \"https://machinelearningmastery.com/wp-content/uploads/2021/08/attention_research_1.png\" width = 400>"]},{"cell_type":"markdown","id":"bb1a97d7","metadata":{"papermill":{"duration":0.011892,"end_time":"2023-07-06T16:04:05.772783","exception":false,"start_time":"2023-07-06T16:04:05.760891","status":"completed"},"tags":[]},"source":["# 2 | Data 📊\n","\n","The `goal` of this competition is to `predict the function of a set of proteins`. We will `develop a model trained` on the `amino-acid sequences` of the `proteins and on other data`. Our work `will help researchers` better `understand the function of proteins`, which is `important for discovering` `how cells, tissues, and organs work`.\n"]},{"cell_type":"code","execution_count":1,"id":"fc835bad","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:05.799823Z","iopub.status.busy":"2023-07-06T16:04:05.799285Z","iopub.status.idle":"2023-07-06T16:04:05.891021Z","shell.execute_reply":"2023-07-06T16:04:05.889669Z"},"papermill":{"duration":0.108832,"end_time":"2023-07-06T16:04:05.894067","exception":false,"start_time":"2023-07-06T16:04:05.785235","status":"completed"},"tags":[]},"outputs":[],"source":["import pandas as pd \n","import re\n","from Bio.Seq import Seq"]},{"cell_type":"markdown","id":"e91dd5b3","metadata":{"papermill":{"duration":0.01187,"end_time":"2023-07-06T16:04:05.918362","exception":false,"start_time":"2023-07-06T16:04:05.906492","status":"completed"},"tags":[]},"source":["The $Training$ $Set$ contains all `proteins with annotated terms` that have been validated by \n","* $Experimental$\n","* $High-Throughput$ $Evidence$\n","* [$Traceable$ $Author$ $Statement$](https://wiki.geneontology.org/index.php/Traceable_Author_Statement_(TAS)#:~:text=The%20TAS%20evidence%20code%20covers,annotations%20come%20from%20review%20articles.)\n","* [$Inferred$ $by$ $Curator$ $(IC)$](https://wiki.geneontology.org/Inferred_by_Curator_(IC)) \n","\n","**Any other sources of Data are allowed**\n","\n","### 2.1.1.1 | Go-Basic.obo\n","\n","The $Ontology$ data is in the `file go-basic.obo`. This file is in $OBO$ `Biology-Oriented Language`. The nodes in `the graph are indexed` by the `term name`\n","```\n","subontology_roots = {'BPO':'GO:0008150',\n","                     'CCO':'GO:0005575',\n","                     'MFO':'GO:0003674'}\n","```"]},{"cell_type":"code","execution_count":2,"id":"378eb87c","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:05.945445Z","iopub.status.busy":"2023-07-06T16:04:05.94461Z","iopub.status.idle":"2023-07-06T16:04:10.409888Z","shell.execute_reply":"2023-07-06T16:04:10.408537Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":4.482153,"end_time":"2023-07-06T16:04:10.412745","exception":false,"start_time":"2023-07-06T16:04:05.930592","status":"completed"},"tags":[]},"outputs":[{"name":"stdout","output_type":"stream","text":["[Term]\n","id: GO:0000001\n","name: mitochondrion inheritance\n","namespace: biological_process\n","def: \"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:11389764]\n","synonym: \"mitochondrial inheritance\" EXACT []\n","is_a: GO:0048308 ! organelle inheritance\n","is_a: GO:0048311 ! mitochondrion distribution\n","\n"]}],"source":["with open('/kaggle/input/cafa-5-protein-function-prediction/Train/go-basic.obo') as file :\n","    \n","    content = file.read()\n","    stanzas =  re.findall(r'\\[Term\\][\\s\\S]*?(?=\\n\\[|$)' , content)\n","    \n","print(stanzas[0])"]},{"cell_type":"markdown","id":"d7e5e456","metadata":{"papermill":{"duration":0.012649,"end_time":"2023-07-06T16:04:10.438175","exception":false,"start_time":"2023-07-06T16:04:10.425526","status":"completed"},"tags":[]},"source":["### 2.1.1.2 | Training Sequences.fasta\n","\n","This file contains only `sequences` for `proteins` with `annotations` in the dataset `labeled proteins`.\n","\n","This files are in `FASTA` format. \n","\n","This file contains . To obtain the full set of protein sequences for unlabeled proteins, the Swiss-Prot and TrEMBL databases can be found here."]},{"cell_type":"code","execution_count":3,"id":"5a910ff2","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:10.464975Z","iopub.status.busy":"2023-07-06T16:04:10.464567Z","iopub.status.idle":"2023-07-06T16:04:11.568267Z","shell.execute_reply":"2023-07-06T16:04:11.567008Z"},"papermill":{"duration":1.119829,"end_time":"2023-07-06T16:04:11.570686","exception":false,"start_time":"2023-07-06T16:04:10.450857","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["'>P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosylase OS=Vaccinia virus (strain Copenhagen) OX=10249 GN=UNG PE=1 SV=1\\nMNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIPDKFFIQLKQPLRNK\\nRVCVCGIDPYPKDGTGVPFESPNF'"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["with open(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_sequences.fasta\" , \"r\") as file:\n","    x = file.read()\n","    \n","x[:200]"]},{"cell_type":"markdown","id":"785c8ed9","metadata":{"papermill":{"duration":0.012224,"end_time":"2023-07-06T16:04:11.595507","exception":false,"start_time":"2023-07-06T16:04:11.583283","status":"completed"},"tags":[]},"source":["### 2.1.1.3 | Train Terms.tsv\n","\n","This file contains the list of annotated terms `ground truth` for the proteins in `train_sequences.fasta`. "]},{"cell_type":"code","execution_count":4,"id":"cdc7e494","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:11.622394Z","iopub.status.busy":"2023-07-06T16:04:11.621957Z","iopub.status.idle":"2023-07-06T16:04:15.784503Z","shell.execute_reply":"2023-07-06T16:04:15.783086Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":4.179483,"end_time":"2023-07-06T16:04:15.78741","exception":false,"start_time":"2023-07-06T16:04:11.607927","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>5363858</th>\n","      <td>X5L565</td>\n","      <td>GO:0050649</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363859</th>\n","      <td>X5L565</td>\n","      <td>GO:0016491</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363860</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005515</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363861</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005488</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363862</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0003674</td>\n","      <td>MFO</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5363863 rows × 3 columns</p>\n","</div>"],"text/plain":["            EntryID        term aspect\n","0        A0A009IHW8  GO:0008152    BPO\n","1        A0A009IHW8  GO:0034655    BPO\n","2        A0A009IHW8  GO:0072523    BPO\n","3        A0A009IHW8  GO:0044270    BPO\n","4        A0A009IHW8  GO:0006753    BPO\n","...             ...         ...    ...\n","5363858      X5L565  GO:0050649    MFO\n","5363859      X5L565  GO:0016491    MFO\n","5363860      X5M5N0  GO:0005515    MFO\n","5363861      X5M5N0  GO:0005488    MFO\n","5363862      X5M5N0  GO:0003674    MFO\n","\n","[5363863 rows x 3 columns]"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_terms.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"53b00f2e","metadata":{"papermill":{"duration":0.01244,"end_time":"2023-07-06T16:04:15.812932","exception":false,"start_time":"2023-07-06T16:04:15.800492","status":"completed"},"tags":[]},"source":["The first column indicates the `protein's UniProt accession ID`, the second is the `GO term ID`, and the third indicates in which `ontology the term appears`.\n","\n","### 2.1.1.4 | Train Taxonomy.tsv\n","\n","This file contains the list of `proteins and the species to which they belong`, represented by a `taxonomic identifier` `taxon ID` number."]},{"cell_type":"code","execution_count":5,"id":"909f90ae","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:15.841494Z","iopub.status.busy":"2023-07-06T16:04:15.840168Z","iopub.status.idle":"2023-07-06T16:04:15.976316Z","shell.execute_reply":"2023-07-06T16:04:15.975082Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":0.153067,"end_time":"2023-07-06T16:04:15.979195","exception":false,"start_time":"2023-07-06T16:04:15.826128","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Q8IXT2</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Q04418</td>\n","      <td>559292</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A8DYA3</td>\n","      <td>7227</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Q9UUI3</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q57ZS4</td>\n","      <td>185431</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>142241</th>\n","      <td>Q5TD07</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>142242</th>\n","      <td>A8BB17</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142243</th>\n","      <td>A0A2R8QBB1</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142244</th>\n","      <td>P0CT72</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>142245</th>\n","      <td>Q9NZ43</td>\n","      <td>9606</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>142246 rows × 2 columns</p>\n","</div>"],"text/plain":["           EntryID  taxonomyID\n","0           Q8IXT2        9606\n","1           Q04418      559292\n","2           A8DYA3        7227\n","3           Q9UUI3      284812\n","4           Q57ZS4      185431\n","...            ...         ...\n","142241      Q5TD07        9606\n","142242      A8BB17        7955\n","142243  A0A2R8QBB1        7955\n","142244      P0CT72      284812\n","142245      Q9NZ43        9606\n","\n","[142246 rows x 2 columns]"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_taxonomy.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"c9b94a78","metadata":{"papermill":{"duration":0.013375,"end_time":"2023-07-06T16:04:16.005865","exception":false,"start_time":"2023-07-06T16:04:15.99249","status":"completed"},"tags":[]},"source":["### 2.1.1.5 | IA.txt\n","\n","IA.txt contains the information accretion (weights) for each GO term. These weights are used to compute weighted precision and recall, as described in the Evaluation section. \n","\n","## 2.1.2 | Test Set\n","\n","The $Test$ $Set$ is `unknown at the beginning` of the competition. It will contain `protein sequences` `their functions` from the `test superset` that `gained experimental annotations` between the `submission-deadline` and the `time of evaluation`.\n","\n","# 3 | PyTorch DataLoader ⚙️"]},{"cell_type":"code","execution_count":6,"id":"9eafbebf","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:16.034848Z","iopub.status.busy":"2023-07-06T16:04:16.034449Z","iopub.status.idle":"2023-07-06T16:04:19.773789Z","shell.execute_reply":"2023-07-06T16:04:19.772813Z"},"papermill":{"duration":3.756353,"end_time":"2023-07-06T16:04:19.776442","exception":false,"start_time":"2023-07-06T16:04:16.020089","status":"completed"},"tags":[]},"outputs":[],"source":["import numpy as np\n","\n","import torch\n","from torch.utils.data import Dataset"]},{"cell_type":"markdown","id":"bcc2797a","metadata":{"papermill":{"duration":0.012856,"end_time":"2023-07-06T16:04:19.802922","exception":false,"start_time":"2023-07-06T16:04:19.790066","status":"completed"},"tags":[]},"source":["The dataloader is highly inspired by **[Henri Upton](https://www.kaggle.com/henriupton)=>[ProteiNet 🧬 PyTorch+EMS2/T5/ProtBERT Embeddings](https://www.kaggle.com/code/henriupton/proteinet-pytorch-ems2-t5-protbert-embeddings)**\n","\n","First we will make a simple class..."]},{"cell_type":"code","execution_count":7,"id":"008e3cd2","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:19.830965Z","iopub.status.busy":"2023-07-06T16:04:19.830322Z","iopub.status.idle":"2023-07-06T16:04:19.835547Z","shell.execute_reply":"2023-07-06T16:04:19.834508Z"},"papermill":{"duration":0.021864,"end_time":"2023-07-06T16:04:19.837798","exception":false,"start_time":"2023-07-06T16:04:19.815934","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    pass"]},{"cell_type":"markdown","id":"a2341f80","metadata":{"papermill":{"duration":0.012738,"end_time":"2023-07-06T16:04:19.86366","exception":false,"start_time":"2023-07-06T16:04:19.850922","status":"completed"},"tags":[]},"source":["Now we will load the `embeds` from the `T5 Embeds`\n","\n","Embeds are like representation of a non-numercial elemenet to a list of numerical elemenet. "]},{"cell_type":"code","execution_count":8,"id":"ca82ca07","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:19.891644Z","iopub.status.busy":"2023-07-06T16:04:19.890952Z","iopub.status.idle":"2023-07-06T16:04:19.89606Z","shell.execute_reply":"2023-07-06T16:04:19.8953Z"},"papermill":{"duration":0.021941,"end_time":"2023-07-06T16:04:19.898566","exception":false,"start_time":"2023-07-06T16:04:19.876625","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    \n","    def __init__(self):\n","        \n","        super(P_Dataset).__init__()\n","        \n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")"]},{"cell_type":"markdown","id":"b38bc843","metadata":{"papermill":{"duration":0.013398,"end_time":"2023-07-06T16:04:19.925515","exception":false,"start_time":"2023-07-06T16:04:19.912117","status":"completed"},"tags":[]},"source":["In case you wanna see, what these `embeds`looks like, open the below hidden cells"]},{"cell_type":"code","execution_count":9,"id":"191dd1c5","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:19.953871Z","iopub.status.busy":"2023-07-06T16:04:19.9532Z","iopub.status.idle":"2023-07-06T16:04:31.562169Z","shell.execute_reply":"2023-07-06T16:04:31.560882Z"},"papermill":{"duration":11.626798,"end_time":"2023-07-06T16:04:31.565473","exception":false,"start_time":"2023-07-06T16:04:19.938675","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["(array([[ 0.05470492,  0.06342026, -0.01531996, ..., -0.04331931,\n","          0.03600927,  0.06309301],\n","        [ 0.09037268,  0.08984205, -0.02388695, ..., -0.05335043,\n","          0.01964429,  0.07962959],\n","        [ 0.04358805,  0.03957234, -0.01433173, ..., -0.04446448,\n","          0.03097377,  0.04032155],\n","        ...,\n","        [ 0.03274843,  0.14186755,  0.03414193, ...,  0.0206462 ,\n","          0.05467706, -0.01504807],\n","        [ 0.05271317,  0.15701264,  0.04327936, ...,  0.02159062,\n","          0.0625835 , -0.01490347],\n","        [ 0.04187706,  0.13404095,  0.0790938 , ...,  0.0541121 ,\n","          0.01576663, -0.02881737]]),\n"," (141865, 1024))"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","\n","embeds , embeds.shape"]},{"cell_type":"markdown","id":"0313973d","metadata":{"papermill":{"duration":0.013031,"end_time":"2023-07-06T16:04:31.592676","exception":false,"start_time":"2023-07-06T16:04:31.579645","status":"completed"},"tags":[]},"source":["For the `training` we need the `targets`, which we will take from the `train_targets_top500.npy`."]},{"cell_type":"code","execution_count":10,"id":"13608400","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:31.621746Z","iopub.status.busy":"2023-07-06T16:04:31.620814Z","iopub.status.idle":"2023-07-06T16:04:31.628324Z","shell.execute_reply":"2023-07-06T16:04:31.627139Z"},"papermill":{"duration":0.024823,"end_time":"2023-07-06T16:04:31.630904","exception":false,"start_time":"2023-07-06T16:04:31.606081","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    \n","    def __init__(self , datatype):\n","        \n","        super(P_Dataset).__init__()\n","        \n","        self.datatype  = datatype\n","        \n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","        \n","        if self.datatype == \"train\" : \n","            \n","            self.targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:self.embeds.shape[0]]"]},{"cell_type":"markdown","id":"06ae7c69","metadata":{"papermill":{"duration":0.013244,"end_time":"2023-07-06T16:04:31.657813","exception":false,"start_time":"2023-07-06T16:04:31.644569","status":"completed"},"tags":[]},"source":["These are our targets"]},{"cell_type":"code","execution_count":11,"id":"618041df","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:31.686961Z","iopub.status.busy":"2023-07-06T16:04:31.686479Z","iopub.status.idle":"2023-07-06T16:04:35.77278Z","shell.execute_reply":"2023-07-06T16:04:35.771702Z"},"papermill":{"duration":4.104226,"end_time":"2023-07-06T16:04:35.77541","exception":false,"start_time":"2023-07-06T16:04:31.671184","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["array([[0., 1., 0., ..., 0., 0., 0.],\n","       [1., 1., 1., ..., 0., 1., 1.],\n","       [1., 1., 1., ..., 0., 0., 0.],\n","       ...,\n","       [1., 0., 1., ..., 0., 0., 0.],\n","       [0., 1., 0., ..., 0., 0., 0.],\n","       [1., 0., 1., ..., 0., 0., 0.]])"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:embeds.shape[0]]\n","\n","targets"]},{"cell_type":"markdown","id":"d999bde3","metadata":{"papermill":{"duration":0.012979,"end_time":"2023-07-06T16:04:35.802501","exception":false,"start_time":"2023-07-06T16:04:35.789522","status":"completed"},"tags":[]},"source":["Now we will just apply some `Getters` in the class "]},{"cell_type":"code","execution_count":12,"id":"dd56f367","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:35.831392Z","iopub.status.busy":"2023-07-06T16:04:35.830985Z","iopub.status.idle":"2023-07-06T16:04:35.840448Z","shell.execute_reply":"2023-07-06T16:04:35.839126Z"},"papermill":{"duration":0.026748,"end_time":"2023-07-06T16:04:35.842866","exception":false,"start_time":"2023-07-06T16:04:35.816118","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","\n","    def __init__(self , datatype ):\n","        super(P_Dataset).__init__()\n","\n","        self.datatype = datatype\n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","        \n","        if datatype == \"train\":\n","\n","            self.targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:self.embeds.shape[0]]\n","            \n","    def __len__(self):return self.targets.shape[0]\n","\n","    def __getitem__(self , index):\n","\n","        r_embed = torch.tensor(self.embeds[index] , dtype = torch.float32)\n","\n","        if self.datatype == \"train\":\n","\n","            r_targets = torch.tensor(self.targets[index], dtype = torch.float32)\n","\n","            return r_embed, r_targets\n","\n","        return r_embed"]},{"cell_type":"code","execution_count":13,"id":"b0586f2f","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:35.873403Z","iopub.status.busy":"2023-07-06T16:04:35.872176Z","iopub.status.idle":"2023-07-06T16:04:36.687553Z","shell.execute_reply":"2023-07-06T16:04:36.686118Z"},"papermill":{"duration":0.834418,"end_time":"2023-07-06T16:04:36.690858","exception":false,"start_time":"2023-07-06T16:04:35.85644","status":"completed"},"tags":[]},"outputs":[],"source":["train_data = P_Dataset(datatype = \"train\")"]},{"cell_type":"code","execution_count":14,"id":"9cd0f408","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:36.721066Z","iopub.status.busy":"2023-07-06T16:04:36.720636Z","iopub.status.idle":"2023-07-06T16:04:36.72706Z","shell.execute_reply":"2023-07-06T16:04:36.726144Z"},"papermill":{"duration":0.024027,"end_time":"2023-07-06T16:04:36.729498","exception":false,"start_time":"2023-07-06T16:04:36.705471","status":"completed"},"tags":[]},"outputs":[],"source":["train_dataloader = torch.utils.data.DataLoader(train_data , batch_size = 128 , shuffle = True)"]},{"cell_type":"markdown","id":"f0c562d5","metadata":{"papermill":{"duration":0.013239,"end_time":"2023-07-06T16:04:36.756719","exception":false,"start_time":"2023-07-06T16:04:36.74348","status":"completed"},"tags":[]},"source":["# 4 | Model Setup \n","\n","Now we will train Simple models, and see the results\n","\n","# 4.1 | MultiLayerPerceptron \n","\n","A `MultiLayerPerceptron` is a simple model, which consists of just few perceptrons/nodes/neurons connected with each other. \n","\n","As we have choosen for the `top-500` targets, our last layer of the network will contain 500 perceptrons, "]},{"cell_type":"code","execution_count":15,"id":"98a6abd7","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:36.786128Z","iopub.status.busy":"2023-07-06T16:04:36.785321Z","iopub.status.idle":"2023-07-06T16:04:36.797591Z","shell.execute_reply":"2023-07-06T16:04:36.796086Z"},"papermill":{"duration":0.030337,"end_time":"2023-07-06T16:04:36.80075","exception":false,"start_time":"2023-07-06T16:04:36.770413","status":"completed"},"tags":[]},"outputs":[],"source":["class MultiLayerPerceptron(torch.nn.Module):\n","\n","    def __init__(self):\n","      \n","        super(MultiLayerPerceptron, self).__init__()\n","\n","        # l_1 = Linear Layer 1 \n","        # a_1 = Activation Layer 1\n","\n","        self.l_1 = torch.nn.Linear(1024 , 1000)\n","        self.a_1 = torch.nn.ReLU()\n","\n","        self.l_2 = torch.nn.Linear(1000 , 900)\n","        self.a_2 = torch.nn.ReLU()\n","        \n","        self.l_3 = torch.nn.Linear(900 , 800)\n","        self.a_3 = torch.nn.ReLU()\n","        \n","        self.l_4 = torch.nn.Linear(800 , 700)\n","        self.a_4 = torch.nn.ReLU()\n","        \n","        self.l_5 = torch.nn.Linear(700 , 600)\n","        self.a_5 = torch.nn.ReLU()\n","        \n","        self.l_6 = torch.nn.Linear(600 , 500)\n","    \n","    def forward(self, x):\n","        \n","        x = self.l_1(x)\n","        x = self.a_1(x)\n","\n","        x = self.l_2(x)\n","        x = self.a_2(x)\n","        \n","        x = self.l_3(x)\n","        x = self.a_3(x)\n","        \n","        x = self.l_4(x)\n","        x = self.a_4(x)\n","        \n","        x = self.l_5(x)\n","        x = self.a_5(x)\n","        \n","        x = self.l_6(x)\n","        \n","        return x"]},{"cell_type":"markdown","id":"cfbf047a","metadata":{"papermill":{"duration":0.013046,"end_time":"2023-07-06T16:04:36.827895","exception":false,"start_time":"2023-07-06T16:04:36.814849","status":"completed"},"tags":[]},"source":["I cannot use `Kaggle GPUs` for some reason (I dont know why), so I will be commenting the `cuda` parts "]},{"cell_type":"code","execution_count":16,"id":"8e209715","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:36.856845Z","iopub.status.busy":"2023-07-06T16:04:36.855998Z","iopub.status.idle":"2023-07-06T16:04:36.944185Z","shell.execute_reply":"2023-07-06T16:04:36.942994Z"},"papermill":{"duration":0.105716,"end_time":"2023-07-06T16:04:36.946956","exception":false,"start_time":"2023-07-06T16:04:36.84124","status":"completed"},"tags":[]},"outputs":[],"source":["MLP = MultiLayerPerceptron()\n","\n","# MLP = MultiLayerPerceptron().to(\"cuda\")"]},{"cell_type":"markdown","id":"e486cd96","metadata":{"papermill":{"duration":0.013054,"end_time":"2023-07-06T16:04:36.973585","exception":false,"start_time":"2023-07-06T16:04:36.960531","status":"completed"},"tags":[]},"source":["# 5 | Loss Function \n","\n","$Loss$ is a `measure` of `how well a model is performing` on a given task. It is `calculated` by `comparing` the `model's predictions` to the `ground truth labels`. The `lower the loss`, the `better the model` is performing.\n","\n","Here we will be using `Cross Entropy Loss` , a loss function that `measures` the `difference` between the `model's predicted probability distribution` and the `ground truth distribution`."]},{"cell_type":"code","execution_count":17,"id":"c900a8a5","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:37.002964Z","iopub.status.busy":"2023-07-06T16:04:37.002552Z","iopub.status.idle":"2023-07-06T16:04:37.00923Z","shell.execute_reply":"2023-07-06T16:04:37.007839Z"},"papermill":{"duration":0.025181,"end_time":"2023-07-06T16:04:37.012192","exception":false,"start_time":"2023-07-06T16:04:36.987011","status":"completed"},"tags":[]},"outputs":[],"source":["CrossEntropy = torch.nn.CrossEntropyLoss()"]},{"cell_type":"markdown","id":"65beffc0","metadata":{"papermill":{"duration":0.012965,"end_time":"2023-07-06T16:04:37.039081","exception":false,"start_time":"2023-07-06T16:04:37.026116","status":"completed"},"tags":[]},"source":["# 6 | Optimizer \n","\n","An $Optimizer$ is an `algorithm` or function that `updates the weights and biases` of a neural network in order to `minimize a loss function`. \n","\n","Here we will be using the `Adam Optimizer`\n","\n","$Adam$ is an $Adaptive$ $Learning$ $Rate$ method, which works by `maintaining two moving averages of the gradients`\n","* $Mean$ - Calculate the `momentum term`, which helps to `prevent` the `optimizer` from `getting stuck in local minima`\n","* $Variance$ - Calculate the `learning rate`, which is `adjusted based on the magnitude of the gradients`.\n","\n","$$m_t = \\beta_1 * m_{t - 1} + (1 - \\beta_1) * w_t$$\n","\n","$$v_t = \\beta_2 * m_{t - 1} + (1 - \\beta_2) * w_t$$\n","\n","$$m_t = \\frac{m_t}{1 - \\beta_1^t}$$\n","\n","$$v_t = \\frac{v_t}{1 - \\beta_2^t}$$\n","\n","$$w_{t+1} = w_t - \\frac{n}{\\sqrt{v_t + e}} * m_t$$"]},{"cell_type":"code","execution_count":18,"id":"d73ac6c8","metadata":{"execution":{"iopub.execute_input":"2023-07-06T16:04:37.068218Z","iopub.status.busy":"2023-07-06T16:04:37.067778Z","iopub.status.idle":"2023-07-06T16:04:37.073189Z","shell.execute_reply":"2023-07-06T16:04:37.071721Z"},"papermill":{"duration":0.022367,"end_time":"2023-07-06T16:04:37.07544","exception":false,"start_time":"2023-07-06T16:04:37.053073","status":"completed"},"tags":[]},"outputs":[],"source":["optimizer = torch.optim.Adam(MLP.parameters(), lr = 0.0005)"]},{"cell_type":"markdown","id":"1ad4cda7","metadata":{"papermill":{"duration":0.013172,"end_time":"2023-07-06T16:04:37.102179","exception":false,"start_time":"2023-07-06T16:04:37.089007","status":"completed"},"tags":[]},"source":["# 7 | Training Loop  "]},{"cell_type":"code","execution_count":19,"id":"fa3a784c","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:37.131122Z","iopub.status.busy":"2023-07-06T16:04:37.130677Z","iopub.status.idle":"2023-07-06T16:04:37.135961Z","shell.execute_reply":"2023-07-06T16:04:37.134767Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":0.022647,"end_time":"2023-07-06T16:04:37.138414","exception":false,"start_time":"2023-07-06T16:04:37.115767","status":"completed"},"tags":[]},"outputs":[],"source":["from IPython.display import IFrame"]},{"cell_type":"markdown","id":"deb59090","metadata":{"papermill":{"duration":0.013027,"end_time":"2023-07-06T16:04:37.165295","exception":false,"start_time":"2023-07-06T16:04:37.152268","status":"completed"},"tags":[]},"source":["Now we will start the training loop \n","\n","I dont the exact reason, but everytime I try to access `GPU` for some training in `Kaggle`. `CUDA goes out of memory`. Thus I have trained the model on `Colab` and will imported the results to `Wandb`."]},{"cell_type":"markdown","id":"41b21747","metadata":{"papermill":{"duration":0.012998,"end_time":"2023-07-06T16:04:37.191699","exception":false,"start_time":"2023-07-06T16:04:37.178701","status":"completed"},"tags":[]},"source":["```\n","los = []\n","\n","for epochs in (range(5)):\n","\n","    losses = []\n","\n","    for x , y in tqdm.tqdm(train_dataloader):\n","        \n","        x = torch.tensor(x , dtype = torch.float32)\n","        y = torch.tensor(x , dtype = troch.float32)\n","        \n","        \n","#         x = torch.tensor(x , dtype = torch.float32).to(\"cuda\")\n","#         y = torch.tensor(y , dtype = torch.float32).to(\"cuda\")\n","\n","        optimizer.zero_grad()\n","        preds = model(x)\n","\n","        loss = CrossEntropy(preds, y)\n","        losses.append(loss)\n","\n","    los.append(losses)\n","```"]},{"cell_type":"markdown","id":"34baea0b","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"papermill":{"duration":0.01295,"end_time":"2023-07-06T16:04:37.218215","exception":false,"start_time":"2023-07-06T16:04:37.205265","status":"completed"},"tags":[]},"source":["```\n","---------------------------------------------------------------------------\n","OutOfMemoryError                          Traceback (most recent call last)\n","Cell In[31], line 15\n","     12 img = img.to(\"cuda\")\n","     13 mask = mask.to(\"cuda\")\n","---> 15 outputs = model(img)  \n","     17 loss =  loss_func(outputs , mask)\n","     19 loss.backward()\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/segmentation_models_pytorch/base/model.py:29, in SegmentationModel.forward(self, x)\n","     25 \"\"\"Sequentially pass `x` trough model`s encoder, decoder and heads\"\"\"\n","     27 self.check_input_shape(x)\n","---> 29 features = self.encoder(x)\n","     30 decoder_output = self.decoder(*features)\n","     32 masks = self.segmentation_head(decoder_output)\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/segmentation_models_pytorch/encoders/efficientnet.py:73, in EfficientNetEncoder.forward(self, x)\n","     71             drop_connect = drop_connect_rate * block_number / len(self._blocks)\n","     72             block_number += 1.0\n","---> 73             x = module(x, drop_connect)\n","     75     features.append(x)\n","     77 return features\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/model.py:111, in MBConvBlock.forward(self, inputs, drop_connect_rate)\n","    109 x = self._depthwise_conv(x)\n","    110 x = self._bn1(x)\n","--> 111 x = self._swish(x)\n","    113 # Squeeze and Excitation\n","    114 if self.has_se:\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/utils.py:80, in MemoryEfficientSwish.forward(self, x)\n","     79 def forward(self, x):\n","---> 80     return SwishImplementation.apply(x)\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/autograd/function.py:506, in Function.apply(cls, *args, **kwargs)\n","    503 if not torch._C._are_functorch_transforms_active():\n","    504     # See NOTE: [functorch vjp and autograd interaction]\n","    505     args = _functorch.utils.unwrap_dead_wrappers(args)\n","--> 506     return super().apply(*args, **kwargs)  # type: ignore[misc]\n","    508 if cls.setup_context == _SingleLevelFunction.setup_context:\n","    509     raise RuntimeError(\n","    510         'In order to use an autograd.Function with functorch transforms '\n","    511         '(vmap, grad, jvp, jacrev, ...), it must override the setup_context '\n","    512         'staticmethod. For more details, please see '\n","    513         'https://pytorch.org/docs/master/notes/extending.func.html style=\"color:rgb(175,0,0)\">')\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/utils.py:67, in SwishImplementation.forward(ctx, i)\n","     65 @staticmethod\n","     66 def forward(ctx, i):\n","---> 67     result = i * torch.sigmoid(i)\n","     68     ctx.save_for_backward(i)\n","     69     return result\n","\n","OutOfMemoryError: CUDA out of memory. Tried to allocate 40.00 MiB (GPU 0; 15.90 GiB total capacity; 319.37 MiB already allocated; 7.75 MiB free; 326.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF\n","```"]},{"cell_type":"code","execution_count":20,"id":"23691143","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-06T16:04:37.246707Z","iopub.status.busy":"2023-07-06T16:04:37.24627Z","iopub.status.idle":"2023-07-06T16:04:37.254342Z","shell.execute_reply":"2023-07-06T16:04:37.253172Z"},"papermill":{"duration":0.025318,"end_time":"2023-07-06T16:04:37.256862","exception":false,"start_time":"2023-07-06T16:04:37.231544","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["\n","        <iframe\n","            width=\"1300\"\n","            height=\"400\"\n","            src=\"https://wandb.ai//ayushsinghal659/CAFA%7CMLP/reports/CAFA-MLP--Vmlldzo0Nzk1NjQx\"\n","            frameborder=\"0\"\n","            allowfullscreen\n","            \n","        ></iframe>\n","        "],"text/plain":["<IPython.lib.display.IFrame at 0x7c2da2e037f0>"]},"execution_count":20,"metadata":{},"output_type":"execute_result"}],"source":["IFrame(\"https://wandb.ai//ayushsinghal659/CAFA%7CMLP/reports/CAFA-MLP--Vmlldzo0Nzk1NjQx\" , 1300 , 400)"]},{"cell_type":"markdown","id":"f24d7126","metadata":{"papermill":{"duration":0.013262,"end_time":"2023-07-06T16:04:37.284124","exception":false,"start_time":"2023-07-06T16:04:37.270862","status":"completed"},"tags":[]},"source":["As we can see we did not get good results, but we will try to omprove our results, by improving the model and by adding new one "]},{"cell_type":"markdown","id":"beb6a1c1","metadata":{"papermill":{"duration":0.012998,"end_time":"2023-07-06T16:04:37.310558","exception":false,"start_time":"2023-07-06T16:04:37.29756","status":"completed"},"tags":[]},"source":["# 8 | TO DO LIST 📄\n","\n","```\n","TO DO 1 : VISUALIZE THE DATA\n","\n","TO DO 2 : TRAIN A MODEL\n","\n","TO DO 3 : TRY DIFFERENT MODELS\n","\n","TO DO 4 : ADD WANDB SUPPORT\n","\n","TO DO 5 : ADD TENSORFLOW DATA LOADER\n","\n","TO DO 6 : TRAIN A TF MODEL\n","\n","TO DO 7 : IMPROVE RESULTS\n","\n","TO DO 8 : DECREASE TRAINING TIME\n","\n","TO DO 9 : DANCE \n","```\n","\n","# 9 | Ending 🏁\n","\n","**THAT'S IT FOR TODAY GUYS**\n","\n","**WE WILL GO DEEPER INTO THE DATA IN THE UPCOMING VERSIONS**\n","\n","**PLEASE COMMENT YOUR THOUGHTS, HIHGLY APPRICIATED**\n","\n","**DONT FORGET TO MAKE AN UPVOTE, IF YOU LIKED MY WORK $:)$**\n","\n","<img src = \"https://i.imgflip.com/19aadg.jpg\">\n","\n","**PEACE OUT $:)$**"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.10"},"papermill":{"default_parameters":{},"duration":46.787196,"end_time":"2023-07-06T16:04:39.662426","environment_variables":{},"exception":null,"input_path":"__notebook__.ipynb","output_path":"__notebook__.ipynb","parameters":{},"start_time":"2023-07-06T16:03:52.87523","version":"2.4.0"}},"nbformat":4,"nbformat_minor":5}