{"cells":[{"source":"<a href=\"https://www.kaggle.com/code/ayushs9020/utc-training-pytorch-dataloader-cafa?scriptVersionId=135674400\" target=\"_blank\"><img align=\"left\" alt=\"Kaggle\" title=\"Open in Kaggle\" src=\"https://kaggle.com/static/images/open-in-kaggle.svg\"></a>","metadata":{},"cell_type":"markdown"},{"cell_type":"markdown","id":"a40c71f3","metadata":{"papermill":{"duration":0.014162,"end_time":"2023-07-04T07:35:36.303624","exception":false,"start_time":"2023-07-04T07:35:36.289462","status":"completed"},"tags":[]},"source":["# CAFA 5 Protein Prediction\n","\n","<img src = \"https://i.pinimg.com/originals/8e/de/73/8ede737acdcf671e000ae0f87a742e40.png\" width = 400>\n","\n","The `goal` of this competition is to `predict the function of a set of proteins`. We will `develop a model trained` on the `amino-acid sequences` of the `proteins and on other data`. Our work `will help researchers` better `understand the function of proteins`, which is `important for discovering` `how cells, tissues, and organs work`.\n","\n","# 1 | Basic Terminologies 💻\n","\n","* $Structure$ $of$ $a$ $Protein$\n","* $Gene$ $Ontology$ $(GO)$\n","\n","## 1.1 | Structure of a Protein\n","\n","So what is a actually a **Protein...?**\n","\n","First of all lets understand the structure of an **Atom**\n","\n","<img src = \"https://www.sciencefacts.net/wp-content/uploads/2020/11/Parts-of-an-Atom-Diagram.jpg\"  width = 300>\n","\n","There is a really good image I found of the `structure of atom`. Though there are many debates on the structure like this, but this `model is accepted universaly at this moment`.\n","\n","In the centre we have the `Neucleus`. The `Neucleus` is made up of $2$ more structures named as `Neutron` and `Proton`. A `Proton` is `positively charged element` and a `Neutron` is a `neutral charged element`. A `Electron`, `negatively charged element`, `orbits` this `Neucleus` at some `distance apart`.\n","\n","The `more we increase the number` of `Electrons` and `Protons`. The `bigger the atoms becomes`.\n","\n","There are `different shells` where the `Electrons reside`. The `more closer the shell` is, the `less Electrons` it contrains. There are mainly $4$ shells. \n","\n","|||\n","|---|---\n","|$K$|$2$\n","|$L$|$8$\n","|$M$|$18$\n","|$N$|$32$\n","\n","Once an atom `fills its outer most shell` with `Electrons`. It becomes `stable atom` and try to `refuse any donation` or `recieve of extra atom`.\n","\n","<img src = \"https://cdn1.byjus.com/wp-content/uploads/2022/01/word-image128.png\" width = 400>\n","\n","`Different atoms combine` to `share Electrons` and become `stable Molecules` \n","\n","<img src = \"https://www.astrochem.org/sci_img/Amino_Acid_Structure.jpg\" width = 300>\n","\n","A `Amino Acid` is made up of mainly $4$ different atoms\n","`[H , C , O , N]`. \n","\n","<img src = \"https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/L-amino_acid_structure.svg/1200px-L-amino_acid_structure.svg.png\" alt = \"Bro use Light Theme\" width = 300 >\n","\n","We also have a free `Electron Pair` of `Carbon` in this molecule, we call this as a `Side Chain` which can be of different types. Basically this `Side Chain` provide the flexibility to make `different types` of `Amino Acids`. This flexibilty allows for $20$ `different` `Amino Acids` \n","\n","When we join `Amino Acids` with `peptide bonds`, we get `Proteins`. Conncecting different types of `Amino Acids` ends up in different types of `Proteins`.\n","\n","## 1.2 | Gene Ontology (GO)\n","\n","$Gene$ $Ontology$ $(GO)$ is a `controlled vocabulary` that `describes the functions of genes` and gene products. It is constantly being updated as new information becomes available.\n","\n","There are mainly $3$ `Ontologies`\n","\n","||||\n","|---|---|---\n","|$Biological$ $Process$|Describes the `biological processes`|A gene product might be involved in the process of `cell cycle`/`signal transduction.`\n","|$Cellular$ $Component$|Describes the `cellular components`|A gene product might be located in the `nucleus`/`cytoplasm`.\n","|$Molecular$ $Function$|Describes the `molecular functions`|A gene product might be involved in the `catalysis of a reaction`/`binding of a molecule`.\n","\n","**[Gene Ontology Documentation](http://geneontology.org/docs/ontology-documentation/)**"]},{"cell_type":"markdown","id":"0604d65e","metadata":{"papermill":{"duration":0.011572,"end_time":"2023-07-04T07:35:36.327423","exception":false,"start_time":"2023-07-04T07:35:36.315851","status":"completed"},"tags":[]},"source":["# 2 | Data 📊"]},{"cell_type":"code","execution_count":1,"id":"f3f351a9","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:36.35374Z","iopub.status.busy":"2023-07-04T07:35:36.352946Z","iopub.status.idle":"2023-07-04T07:35:36.440497Z","shell.execute_reply":"2023-07-04T07:35:36.439222Z"},"papermill":{"duration":0.104237,"end_time":"2023-07-04T07:35:36.443428","exception":false,"start_time":"2023-07-04T07:35:36.339191","status":"completed"},"tags":[]},"outputs":[],"source":["import pandas as pd \n","import re\n","from Bio.Seq import Seq"]},{"cell_type":"markdown","id":"52940d7a","metadata":{"papermill":{"duration":0.0117,"end_time":"2023-07-04T07:35:36.467319","exception":false,"start_time":"2023-07-04T07:35:36.455619","status":"completed"},"tags":[]},"source":["The $Training$ $Set$ contains all `proteins with annotated terms` that have been validated by \n","* $Experimental$\n","* $High-Throughput$ $Evidence$\n","* [$Traceable$ $Author$ $Statement$](https://wiki.geneontology.org/index.php/Traceable_Author_Statement_(TAS)#:~:text=The%20TAS%20evidence%20code%20covers,annotations%20come%20from%20review%20articles.)\n","* [$Inferred$ $by$ $Curator$ $(IC)$](https://wiki.geneontology.org/Inferred_by_Curator_(IC)) \n","\n","**Any other sources of Data are allowed**\n","\n","### 2.1.1.1 | Go-Basic.obo\n","\n","The $Ontology$ data is in the `file go-basic.obo`. This file is in $OBO$ `Biology-Oriented Language`. The nodes in `the graph are indexed` by the `term name`\n","```\n","subontology_roots = {'BPO':'GO:0008150',\n","                     'CCO':'GO:0005575',\n","                     'MFO':'GO:0003674'}\n","```"]},{"cell_type":"code","execution_count":2,"id":"b91fb69d","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:36.493413Z","iopub.status.busy":"2023-07-04T07:35:36.492743Z","iopub.status.idle":"2023-07-04T07:35:40.917183Z","shell.execute_reply":"2023-07-04T07:35:40.915882Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":4.440625,"end_time":"2023-07-04T07:35:40.920024","exception":false,"start_time":"2023-07-04T07:35:36.479399","status":"completed"},"tags":[]},"outputs":[{"name":"stdout","output_type":"stream","text":["[Term]\n","id: GO:0000001\n","name: mitochondrion inheritance\n","namespace: biological_process\n","def: \"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:11389764]\n","synonym: \"mitochondrial inheritance\" EXACT []\n","is_a: GO:0048308 ! organelle inheritance\n","is_a: GO:0048311 ! mitochondrion distribution\n","\n"]}],"source":["with open('/kaggle/input/cafa-5-protein-function-prediction/Train/go-basic.obo') as file :\n","    \n","    content = file.read()\n","    stanzas =  re.findall(r'\\[Term\\][\\s\\S]*?(?=\\n\\[|$)' , content)\n","    \n","print(stanzas[0])"]},{"cell_type":"markdown","id":"bb52ff18","metadata":{"papermill":{"duration":0.0119,"end_time":"2023-07-04T07:35:40.944145","exception":false,"start_time":"2023-07-04T07:35:40.932245","status":"completed"},"tags":[]},"source":["### 2.1.1.2 | Training Sequences.fasta\n","\n","This file contains only `sequences` for `proteins` with `annotations` in the dataset `labeled proteins`.\n","\n","This files are in `FASTA` format. \n","\n","This file contains . To obtain the full set of protein sequences for unlabeled proteins, the Swiss-Prot and TrEMBL databases can be found here."]},{"cell_type":"code","execution_count":3,"id":"c0a491ba","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:35:40.969894Z","iopub.status.busy":"2023-07-04T07:35:40.969528Z","iopub.status.idle":"2023-07-04T07:35:42.413074Z","shell.execute_reply":"2023-07-04T07:35:42.411979Z"},"papermill":{"duration":1.459033,"end_time":"2023-07-04T07:35:42.415433","exception":false,"start_time":"2023-07-04T07:35:40.9564","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["'>P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosylase OS=Vaccinia virus (strain Copenhagen) OX=10249 GN=UNG PE=1 SV=1\\nMNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIPDKFFIQLKQPLRNK\\nRVCVCGIDPYPKDGTGVPFESPNF'"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["with open(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_sequences.fasta\" , \"r\") as file:\n","    x = file.read()\n","    \n","x[:200]"]},{"cell_type":"markdown","id":"a9827c90","metadata":{"papermill":{"duration":0.013153,"end_time":"2023-07-04T07:35:42.440822","exception":false,"start_time":"2023-07-04T07:35:42.427669","status":"completed"},"tags":[]},"source":["### 2.1.1.3 | Train Terms.tsv\n","\n","This file contains the list of annotated terms `ground truth` for the proteins in `train_sequences.fasta`. "]},{"cell_type":"code","execution_count":4,"id":"23dd8583","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:42.466808Z","iopub.status.busy":"2023-07-04T07:35:42.46639Z","iopub.status.idle":"2023-07-04T07:35:46.470335Z","shell.execute_reply":"2023-07-04T07:35:46.469387Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":4.019719,"end_time":"2023-07-04T07:35:46.472757","exception":false,"start_time":"2023-07-04T07:35:42.453038","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>5363858</th>\n","      <td>X5L565</td>\n","      <td>GO:0050649</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363859</th>\n","      <td>X5L565</td>\n","      <td>GO:0016491</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363860</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005515</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363861</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005488</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363862</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0003674</td>\n","      <td>MFO</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5363863 rows × 3 columns</p>\n","</div>"],"text/plain":["            EntryID        term aspect\n","0        A0A009IHW8  GO:0008152    BPO\n","1        A0A009IHW8  GO:0034655    BPO\n","2        A0A009IHW8  GO:0072523    BPO\n","3        A0A009IHW8  GO:0044270    BPO\n","4        A0A009IHW8  GO:0006753    BPO\n","...             ...         ...    ...\n","5363858      X5L565  GO:0050649    MFO\n","5363859      X5L565  GO:0016491    MFO\n","5363860      X5M5N0  GO:0005515    MFO\n","5363861      X5M5N0  GO:0005488    MFO\n","5363862      X5M5N0  GO:0003674    MFO\n","\n","[5363863 rows x 3 columns]"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_terms.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"57684f47","metadata":{"papermill":{"duration":0.012137,"end_time":"2023-07-04T07:35:46.497604","exception":false,"start_time":"2023-07-04T07:35:46.485467","status":"completed"},"tags":[]},"source":["The first column indicates the `protein's UniProt accession ID`, the second is the `GO term ID`, and the third indicates in which `ontology the term appears`.\n","\n","### 2.1.1.4 | Train Taxonomy.tsv\n","\n","This file contains the list of `proteins and the species to which they belong`, represented by a `taxonomic identifier` `taxon ID` number."]},{"cell_type":"code","execution_count":5,"id":"4aee8421","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:46.525012Z","iopub.status.busy":"2023-07-04T07:35:46.523773Z","iopub.status.idle":"2023-07-04T07:35:46.647799Z","shell.execute_reply":"2023-07-04T07:35:46.646745Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":0.140253,"end_time":"2023-07-04T07:35:46.650344","exception":false,"start_time":"2023-07-04T07:35:46.510091","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Q8IXT2</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Q04418</td>\n","      <td>559292</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A8DYA3</td>\n","      <td>7227</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Q9UUI3</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q57ZS4</td>\n","      <td>185431</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>142241</th>\n","      <td>Q5TD07</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>142242</th>\n","      <td>A8BB17</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142243</th>\n","      <td>A0A2R8QBB1</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142244</th>\n","      <td>P0CT72</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>142245</th>\n","      <td>Q9NZ43</td>\n","      <td>9606</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>142246 rows × 2 columns</p>\n","</div>"],"text/plain":["           EntryID  taxonomyID\n","0           Q8IXT2        9606\n","1           Q04418      559292\n","2           A8DYA3        7227\n","3           Q9UUI3      284812\n","4           Q57ZS4      185431\n","...            ...         ...\n","142241      Q5TD07        9606\n","142242      A8BB17        7955\n","142243  A0A2R8QBB1        7955\n","142244      P0CT72      284812\n","142245      Q9NZ43        9606\n","\n","[142246 rows x 2 columns]"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_taxonomy.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"2eea4933","metadata":{"papermill":{"duration":0.012448,"end_time":"2023-07-04T07:35:46.675805","exception":false,"start_time":"2023-07-04T07:35:46.663357","status":"completed"},"tags":[]},"source":["### 2.1.1.5 | IA.txt\n","\n","IA.txt contains the information accretion (weights) for each GO term. These weights are used to compute weighted precision and recall, as described in the Evaluation section. \n","\n","## 2.1.2 | Test Set\n","\n","The $Test$ $Set$ is `unknown at the beginning` of the competition. It will contain `protein sequences` `their functions` from the `test superset` that `gained experimental annotations` between the `submission-deadline` and the `time of evaluation`.\n","\n","# 3 | PyTorch DataLoader ⚙️"]},{"cell_type":"code","execution_count":6,"id":"102f5a1b","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:46.702155Z","iopub.status.busy":"2023-07-04T07:35:46.701745Z","iopub.status.idle":"2023-07-04T07:35:50.084984Z","shell.execute_reply":"2023-07-04T07:35:50.083966Z"},"papermill":{"duration":3.399342,"end_time":"2023-07-04T07:35:50.087686","exception":false,"start_time":"2023-07-04T07:35:46.688344","status":"completed"},"tags":[]},"outputs":[],"source":["import numpy as np\n","\n","import torch\n","from torch.utils.data import Dataset"]},{"cell_type":"markdown","id":"a14faaad","metadata":{"papermill":{"duration":0.01214,"end_time":"2023-07-04T07:35:50.113012","exception":false,"start_time":"2023-07-04T07:35:50.100872","status":"completed"},"tags":[]},"source":["The dataloader is highly inspired by **[Henri Upton](https://www.kaggle.com/henriupton)=>[ProteiNet 🧬 PyTorch+EMS2/T5/ProtBERT Embeddings](https://www.kaggle.com/code/henriupton/proteinet-pytorch-ems2-t5-protbert-embeddings)**\n","\n","First we will make a simple class..."]},{"cell_type":"code","execution_count":7,"id":"2a312623","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:35:50.140909Z","iopub.status.busy":"2023-07-04T07:35:50.140235Z","iopub.status.idle":"2023-07-04T07:35:50.145317Z","shell.execute_reply":"2023-07-04T07:35:50.144213Z"},"papermill":{"duration":0.022314,"end_time":"2023-07-04T07:35:50.147835","exception":false,"start_time":"2023-07-04T07:35:50.125521","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    pass"]},{"cell_type":"markdown","id":"6016bca6","metadata":{"papermill":{"duration":0.012427,"end_time":"2023-07-04T07:35:50.172993","exception":false,"start_time":"2023-07-04T07:35:50.160566","status":"completed"},"tags":[]},"source":["Now we will load the `embeds` from the `T5 Embeds`\n","\n","Embeds are like representation of a non-numercial elemenet to a list of numerical elemenet. "]},{"cell_type":"code","execution_count":8,"id":"dc4284bb","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:35:50.20009Z","iopub.status.busy":"2023-07-04T07:35:50.199368Z","iopub.status.idle":"2023-07-04T07:35:50.205908Z","shell.execute_reply":"2023-07-04T07:35:50.204821Z"},"papermill":{"duration":0.022507,"end_time":"2023-07-04T07:35:50.208177","exception":false,"start_time":"2023-07-04T07:35:50.18567","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    \n","    def __init__(self):\n","        \n","        super(P_Dataset).__init__()\n","        \n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")"]},{"cell_type":"markdown","id":"40752e07","metadata":{"papermill":{"duration":0.012456,"end_time":"2023-07-04T07:35:50.233759","exception":false,"start_time":"2023-07-04T07:35:50.221303","status":"completed"},"tags":[]},"source":["In case you wanna see, what these `embeds`looks like, open the below hidden cells"]},{"cell_type":"code","execution_count":9,"id":"c1f7d2d8","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-04T07:35:50.261712Z","iopub.status.busy":"2023-07-04T07:35:50.260757Z","iopub.status.idle":"2023-07-04T07:36:00.533577Z","shell.execute_reply":"2023-07-04T07:36:00.532802Z"},"papermill":{"duration":10.288943,"end_time":"2023-07-04T07:36:00.535717","exception":false,"start_time":"2023-07-04T07:35:50.246774","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["(array([[ 0.05470492,  0.06342026, -0.01531996, ..., -0.04331931,\n","          0.03600927,  0.06309301],\n","        [ 0.09037268,  0.08984205, -0.02388695, ..., -0.05335043,\n","          0.01964429,  0.07962959],\n","        [ 0.04358805,  0.03957234, -0.01433173, ..., -0.04446448,\n","          0.03097377,  0.04032155],\n","        ...,\n","        [ 0.03274843,  0.14186755,  0.03414193, ...,  0.0206462 ,\n","          0.05467706, -0.01504807],\n","        [ 0.05271317,  0.15701264,  0.04327936, ...,  0.02159062,\n","          0.0625835 , -0.01490347],\n","        [ 0.04187706,  0.13404095,  0.0790938 , ...,  0.0541121 ,\n","          0.01576663, -0.02881737]]),\n"," (141865, 1024))"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","\n","embeds , embeds.shape"]},{"cell_type":"markdown","id":"b96de4c3","metadata":{"papermill":{"duration":0.012881,"end_time":"2023-07-04T07:36:00.56145","exception":false,"start_time":"2023-07-04T07:36:00.548569","status":"completed"},"tags":[]},"source":["For the `training` we need the `targets`, which we will take from the `train_targets_top500.npy`."]},{"cell_type":"code","execution_count":10,"id":"59dd7789","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:00.589113Z","iopub.status.busy":"2023-07-04T07:36:00.5886Z","iopub.status.idle":"2023-07-04T07:36:00.595017Z","shell.execute_reply":"2023-07-04T07:36:00.593972Z"},"papermill":{"duration":0.022948,"end_time":"2023-07-04T07:36:00.597282","exception":false,"start_time":"2023-07-04T07:36:00.574334","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    \n","    def __init__(self , datatype):\n","        \n","        super(P_Dataset).__init__()\n","        \n","        self.datatype  = datatype\n","        \n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","        \n","        if self.datatype == \"train\" : \n","            \n","            self.targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:self.embeds.shape[0]]"]},{"cell_type":"markdown","id":"11e67ff2","metadata":{"papermill":{"duration":0.012973,"end_time":"2023-07-04T07:36:00.623438","exception":false,"start_time":"2023-07-04T07:36:00.610465","status":"completed"},"tags":[]},"source":["These are our targets"]},{"cell_type":"code","execution_count":11,"id":"0629bb3e","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-04T07:36:00.652752Z","iopub.status.busy":"2023-07-04T07:36:00.652356Z","iopub.status.idle":"2023-07-04T07:36:04.537586Z","shell.execute_reply":"2023-07-04T07:36:04.536592Z"},"papermill":{"duration":3.901858,"end_time":"2023-07-04T07:36:04.540043","exception":false,"start_time":"2023-07-04T07:36:00.638185","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["array([[0., 1., 0., ..., 0., 0., 0.],\n","       [1., 1., 1., ..., 0., 1., 1.],\n","       [1., 1., 1., ..., 0., 0., 0.],\n","       ...,\n","       [1., 0., 1., ..., 0., 0., 0.],\n","       [0., 1., 0., ..., 0., 0., 0.],\n","       [1., 0., 1., ..., 0., 0., 0.]])"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:embeds.shape[0]]\n","\n","targets"]},{"cell_type":"markdown","id":"bc2a6061","metadata":{"papermill":{"duration":0.012823,"end_time":"2023-07-04T07:36:04.566674","exception":false,"start_time":"2023-07-04T07:36:04.553851","status":"completed"},"tags":[]},"source":["Now we will just apply some `Getters` in the class "]},{"cell_type":"code","execution_count":12,"id":"b8982793","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:04.594805Z","iopub.status.busy":"2023-07-04T07:36:04.594399Z","iopub.status.idle":"2023-07-04T07:36:04.603231Z","shell.execute_reply":"2023-07-04T07:36:04.602204Z"},"papermill":{"duration":0.02549,"end_time":"2023-07-04T07:36:04.605481","exception":false,"start_time":"2023-07-04T07:36:04.579991","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","\n","    def __init__(self , datatype ):\n","        super(P_Dataset).__init__()\n","\n","        self.datatype = datatype\n","        self.embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","        \n","        if datatype == \"train\":\n","\n","            self.targets = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:self.embeds.shape[0]]\n","            \n","    def __len__(self):return self.targets.shape[0]\n","\n","    def __getitem__(self , index):\n","\n","        r_embed = torch.tensor(self.embeds[index] , dtype = torch.float32)\n","\n","        if self.datatype == \"train\":\n","\n","            r_targets = torch.tensor(self.targets[index], dtype = torch.float32)\n","\n","            return r_embed, r_targets\n","\n","        return r_embed"]},{"cell_type":"code","execution_count":13,"id":"30c47077","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:04.633445Z","iopub.status.busy":"2023-07-04T07:36:04.632942Z","iopub.status.idle":"2023-07-04T07:36:05.446668Z","shell.execute_reply":"2023-07-04T07:36:05.44548Z"},"papermill":{"duration":0.830867,"end_time":"2023-07-04T07:36:05.449526","exception":false,"start_time":"2023-07-04T07:36:04.618659","status":"completed"},"tags":[]},"outputs":[],"source":["train_data = P_Dataset(datatype = \"train\")"]},{"cell_type":"code","execution_count":14,"id":"9e7386b2","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:05.477758Z","iopub.status.busy":"2023-07-04T07:36:05.477356Z","iopub.status.idle":"2023-07-04T07:36:05.482442Z","shell.execute_reply":"2023-07-04T07:36:05.481404Z"},"papermill":{"duration":0.021998,"end_time":"2023-07-04T07:36:05.484669","exception":false,"start_time":"2023-07-04T07:36:05.462671","status":"completed"},"tags":[]},"outputs":[],"source":["train_dataloader = torch.utils.data.DataLoader(train_data , batch_size = 128 , shuffle = True)"]},{"cell_type":"markdown","id":"d5e4dc17","metadata":{"papermill":{"duration":0.012696,"end_time":"2023-07-04T07:36:05.510478","exception":false,"start_time":"2023-07-04T07:36:05.497782","status":"completed"},"tags":[]},"source":["# 4 | Model Setup \n","\n","Now we will train Simple models, and see the results\n","\n","# 4.1 | MultiLayerPerceptron \n","\n","A `MultiLayerPerceptron` is a simple model, which consists of just few perceptrons/nodes/neurons connected with each other. \n","\n","As we have choosen for the `top-500` targets, our last layer of the network will contain 500 perceptrons, "]},{"cell_type":"code","execution_count":15,"id":"411df42a","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:05.538254Z","iopub.status.busy":"2023-07-04T07:36:05.537814Z","iopub.status.idle":"2023-07-04T07:36:05.549245Z","shell.execute_reply":"2023-07-04T07:36:05.548171Z"},"papermill":{"duration":0.028184,"end_time":"2023-07-04T07:36:05.5517","exception":false,"start_time":"2023-07-04T07:36:05.523516","status":"completed"},"tags":[]},"outputs":[],"source":["class MultiLayerPerceptron(torch.nn.Module):\n","\n","    def __init__(self):\n","      \n","        super(MultiLayerPerceptron, self).__init__()\n","\n","        # l_1 = Linear Layer 1 \n","        # a_1 = Activation Layer 1\n","\n","        self.l_1 = torch.nn.Linear(1024 , 1000)\n","        self.a_1 = torch.nn.ReLU()\n","\n","        self.l_2 = torch.nn.Linear(1000 , 900)\n","        self.a_2 = torch.nn.ReLU()\n","        \n","        self.l_3 = torch.nn.Linear(900 , 800)\n","        self.a_3 = torch.nn.ReLU()\n","        \n","        self.l_4 = torch.nn.Linear(800 , 700)\n","        self.a_4 = torch.nn.ReLU()\n","        \n","        self.l_5 = torch.nn.Linear(700 , 600)\n","        self.a_5 = torch.nn.ReLU()\n","        \n","        self.l_6 = torch.nn.Linear(600 , 500)\n","    \n","    def forward(self, x):\n","        \n","        x = self.l_1(x)\n","        x = self.a_1(x)\n","\n","        x = self.l_2(x)\n","        x = self.a_2(x)\n","        \n","        x = self.l_3(x)\n","        x = self.a_3(x)\n","        \n","        x = self.l_4(x)\n","        x = self.a_4(x)\n","        \n","        x = self.l_5(x)\n","        x = self.a_5(x)\n","        \n","        x = self.l_6(x)\n","        \n","        return x"]},{"cell_type":"markdown","id":"f408f01c","metadata":{"papermill":{"duration":0.0133,"end_time":"2023-07-04T07:36:05.578562","exception":false,"start_time":"2023-07-04T07:36:05.565262","status":"completed"},"tags":[]},"source":["I cannot use `Kaggle GPUs` for some reason (I dont know why), so I will be commenting the `cuda` parts "]},{"cell_type":"code","execution_count":16,"id":"cbaec9cf","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:05.606336Z","iopub.status.busy":"2023-07-04T07:36:05.605915Z","iopub.status.idle":"2023-07-04T07:36:05.691515Z","shell.execute_reply":"2023-07-04T07:36:05.690446Z"},"papermill":{"duration":0.102392,"end_time":"2023-07-04T07:36:05.693978","exception":false,"start_time":"2023-07-04T07:36:05.591586","status":"completed"},"tags":[]},"outputs":[],"source":["MLP = MultiLayerPerceptron()\n","\n","# MLP = MultiLayerPerceptron().to(\"cuda\")"]},{"cell_type":"markdown","id":"c4615ff6","metadata":{"papermill":{"duration":0.012743,"end_time":"2023-07-04T07:36:05.720122","exception":false,"start_time":"2023-07-04T07:36:05.707379","status":"completed"},"tags":[]},"source":["# 5 | Loss Function \n","\n","$Loss$ is a `measure` of `how well a model is performing` on a given task. It is `calculated` by `comparing` the `model's predictions` to the `ground truth labels`. The `lower the loss`, the `better the model` is performing.\n","\n","Here we will be using `Cross Entropy Loss` , a loss function that `measures` the `difference` between the `model's predicted probability distribution` and the `ground truth distribution`."]},{"cell_type":"code","execution_count":17,"id":"79bd5a28","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:05.747565Z","iopub.status.busy":"2023-07-04T07:36:05.747152Z","iopub.status.idle":"2023-07-04T07:36:05.752243Z","shell.execute_reply":"2023-07-04T07:36:05.750977Z"},"papermill":{"duration":0.021503,"end_time":"2023-07-04T07:36:05.754611","exception":false,"start_time":"2023-07-04T07:36:05.733108","status":"completed"},"tags":[]},"outputs":[],"source":["CrossEntropy = torch.nn.CrossEntropyLoss()"]},{"cell_type":"markdown","id":"642b254b","metadata":{"papermill":{"duration":0.012126,"end_time":"2023-07-04T07:36:05.779493","exception":false,"start_time":"2023-07-04T07:36:05.767367","status":"completed"},"tags":[]},"source":["# 6 | Optimizer \n","\n","An $Optimizer$ is an `algorithm` or function that `updates the weights and biases` of a neural network in order to `minimize a loss function`. \n","\n","Here we will be using the `Adam Optimizer`\n","\n","$Adam$ is an $Adaptive$ $Learning$ $Rate$ method, which works by `maintaining two moving averages of the gradients`\n","* $Mean$ - Calculate the `momentum term`, which helps to `prevent` the `optimizer` from `getting stuck in local minima`\n","* $Variance$ - Calculate the `learning rate`, which is `adjusted based on the magnitude of the gradients`.\n","\n","$$m_t = \\beta_1 * m_{t - 1} + (1 - \\beta_1) * w_t$$\n","\n","$$v_t = \\beta_2 * m_{t - 1} + (1 - \\beta_2) * w_t$$\n","\n","$$m_t = \\frac{m_t}{1 - \\beta_1^t}$$\n","\n","$$v_t = \\frac{v_t}{1 - \\beta_2^t}$$\n","\n","$$w_{t+1} = w_t - \\frac{n}{\\sqrt{v_t + e}} * m_t$$"]},{"cell_type":"code","execution_count":18,"id":"410c298b","metadata":{"execution":{"iopub.execute_input":"2023-07-04T07:36:05.806335Z","iopub.status.busy":"2023-07-04T07:36:05.805881Z","iopub.status.idle":"2023-07-04T07:36:05.810961Z","shell.execute_reply":"2023-07-04T07:36:05.809862Z"},"papermill":{"duration":0.021379,"end_time":"2023-07-04T07:36:05.813262","exception":false,"start_time":"2023-07-04T07:36:05.791883","status":"completed"},"tags":[]},"outputs":[],"source":["optimizer = torch.optim.Adam(MLP.parameters(), lr = 0.0005)"]},{"cell_type":"markdown","id":"a65c0dcd","metadata":{"papermill":{"duration":0.012242,"end_time":"2023-07-04T07:36:05.838693","exception":false,"start_time":"2023-07-04T07:36:05.826451","status":"completed"},"tags":[]},"source":["# 7 | Training Loop  "]},{"cell_type":"code","execution_count":19,"id":"00e63612","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:36:05.865435Z","iopub.status.busy":"2023-07-04T07:36:05.865077Z","iopub.status.idle":"2023-07-04T07:36:05.87Z","shell.execute_reply":"2023-07-04T07:36:05.868897Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":0.020831,"end_time":"2023-07-04T07:36:05.872183","exception":false,"start_time":"2023-07-04T07:36:05.851352","status":"completed"},"tags":[]},"outputs":[],"source":["from IPython.display import IFrame"]},{"cell_type":"markdown","id":"fadeb7e2","metadata":{"papermill":{"duration":0.01266,"end_time":"2023-07-04T07:36:05.897844","exception":false,"start_time":"2023-07-04T07:36:05.885184","status":"completed"},"tags":[]},"source":["Now we will start the training loop \n","\n","I dont the exact reason, but everytime I try to access `GPU` for some training in `Kaggle`. `CUDA goes out of memory`. Thus I have trained the model on `Colab` and will imported the results to `Wandb`."]},{"cell_type":"markdown","id":"061fda8a","metadata":{"papermill":{"duration":0.012723,"end_time":"2023-07-04T07:36:05.923753","exception":false,"start_time":"2023-07-04T07:36:05.91103","status":"completed"},"tags":[]},"source":["```\n","los = []\n","\n","for epochs in (range(5)):\n","\n","    losses = []\n","\n","    for x , y in tqdm.tqdm(train_dataloader):\n","        \n","        x = torch.tensor(x , dtype = torch.float32)\n","        y = torch.tensor(x , dtype = troch.float32)\n","        \n","        \n","#         x = torch.tensor(x , dtype = torch.float32).to(\"cuda\")\n","#         y = torch.tensor(y , dtype = torch.float32).to(\"cuda\")\n","\n","        optimizer.zero_grad()\n","        preds = model(x)\n","\n","        loss = CrossEntropy(preds, y)\n","        losses.append(loss)\n","\n","    los.append(losses)\n","```"]},{"cell_type":"markdown","id":"f051c3ac","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"papermill":{"duration":0.012593,"end_time":"2023-07-04T07:36:05.949347","exception":false,"start_time":"2023-07-04T07:36:05.936754","status":"completed"},"tags":[]},"source":["```\n","---------------------------------------------------------------------------\n","OutOfMemoryError                          Traceback (most recent call last)\n","Cell In[31], line 15\n","     12 img = img.to(\"cuda\")\n","     13 mask = mask.to(\"cuda\")\n","---> 15 outputs = model(img)  \n","     17 loss =  loss_func(outputs , mask)\n","     19 loss.backward()\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/segmentation_models_pytorch/base/model.py:29, in SegmentationModel.forward(self, x)\n","     25 \"\"\"Sequentially pass `x` trough model`s encoder, decoder and heads\"\"\"\n","     27 self.check_input_shape(x)\n","---> 29 features = self.encoder(x)\n","     30 decoder_output = self.decoder(*features)\n","     32 masks = self.segmentation_head(decoder_output)\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/segmentation_models_pytorch/encoders/efficientnet.py:73, in EfficientNetEncoder.forward(self, x)\n","     71             drop_connect = drop_connect_rate * block_number / len(self._blocks)\n","     72             block_number += 1.0\n","---> 73             x = module(x, drop_connect)\n","     75     features.append(x)\n","     77 return features\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/model.py:111, in MBConvBlock.forward(self, inputs, drop_connect_rate)\n","    109 x = self._depthwise_conv(x)\n","    110 x = self._bn1(x)\n","--> 111 x = self._swish(x)\n","    113 # Squeeze and Excitation\n","    114 if self.has_se:\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)\n","   1496 # If we don't have any hooks, we want to skip the rest of the logic in\n","   1497 # this function, and just call forward.\n","   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks\n","   1499         or _global_backward_pre_hooks or _global_backward_hooks\n","   1500         or _global_forward_hooks or _global_forward_pre_hooks):\n","-> 1501     return forward_call(*args, **kwargs)\n","   1502 # Do not call functions when jit is used\n","   1503 full_backward_hooks, non_full_backward_hooks = [], []\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/utils.py:80, in MemoryEfficientSwish.forward(self, x)\n","     79 def forward(self, x):\n","---> 80     return SwishImplementation.apply(x)\n","\n","File /opt/conda/lib/python3.10/site-packages/torch/autograd/function.py:506, in Function.apply(cls, *args, **kwargs)\n","    503 if not torch._C._are_functorch_transforms_active():\n","    504     # See NOTE: [functorch vjp and autograd interaction]\n","    505     args = _functorch.utils.unwrap_dead_wrappers(args)\n","--> 506     return super().apply(*args, **kwargs)  # type: ignore[misc]\n","    508 if cls.setup_context == _SingleLevelFunction.setup_context:\n","    509     raise RuntimeError(\n","    510         'In order to use an autograd.Function with functorch transforms '\n","    511         '(vmap, grad, jvp, jacrev, ...), it must override the setup_context '\n","    512         'staticmethod. For more details, please see '\n","    513         'https://pytorch.org/docs/master/notes/extending.func.html style=\"color:rgb(175,0,0)\">')\n","\n","File /opt/conda/lib/python3.10/site-packages/efficientnet_pytorch/utils.py:67, in SwishImplementation.forward(ctx, i)\n","     65 @staticmethod\n","     66 def forward(ctx, i):\n","---> 67     result = i * torch.sigmoid(i)\n","     68     ctx.save_for_backward(i)\n","     69     return result\n","\n","OutOfMemoryError: CUDA out of memory. Tried to allocate 40.00 MiB (GPU 0; 15.90 GiB total capacity; 319.37 MiB already allocated; 7.75 MiB free; 326.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF\n","```"]},{"cell_type":"code","execution_count":20,"id":"d324a282","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-04T07:36:05.977206Z","iopub.status.busy":"2023-07-04T07:36:05.976796Z","iopub.status.idle":"2023-07-04T07:36:05.983906Z","shell.execute_reply":"2023-07-04T07:36:05.98289Z"},"papermill":{"duration":0.023349,"end_time":"2023-07-04T07:36:05.986029","exception":false,"start_time":"2023-07-04T07:36:05.96268","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["\n","        <iframe\n","            width=\"1300\"\n","            height=\"400\"\n","            src=\"https://wandb.ai//ayushsinghal659/CAFA%7CMLP/reports/CAFA-MLP--Vmlldzo0Nzk1NjQx\"\n","            frameborder=\"0\"\n","            allowfullscreen\n","            \n","        ></iframe>\n","        "],"text/plain":["<IPython.lib.display.IFrame at 0x788bd058eb60>"]},"execution_count":20,"metadata":{},"output_type":"execute_result"}],"source":["IFrame(\"https://wandb.ai//ayushsinghal659/CAFA%7CMLP/reports/CAFA-MLP--Vmlldzo0Nzk1NjQx\" , 1300 , 400)"]},{"cell_type":"markdown","id":"9ca206d3","metadata":{"papermill":{"duration":0.012766,"end_time":"2023-07-04T07:36:06.012016","exception":false,"start_time":"2023-07-04T07:36:05.99925","status":"completed"},"tags":[]},"source":["As we can see we did not get good results, but we will try to omprove our results, by improving the model and by adding new one "]},{"cell_type":"markdown","id":"a2b36310","metadata":{"papermill":{"duration":0.012423,"end_time":"2023-07-04T07:36:06.037519","exception":false,"start_time":"2023-07-04T07:36:06.025096","status":"completed"},"tags":[]},"source":["# 8 | TO DO LIST 📄\n","\n","```\n","TO DO 1 : VISUALIZE THE DATA\n","\n","TO DO 2 : TRAIN A MODEL\n","\n","TO DO 3 : TRY DIFFERENT MODELS\n","\n","TO DO 4 : ADD WANDB SUPPORT\n","\n","TO DO 5 : ADD TENSORFLOW DATA LOADER\n","\n","TO DO 6 : TRAIN A TF MODEL\n","\n","TO DO 7 : IMPROVE RESULTS\n","\n","TO DO 8 : DECREASE TRAINING TIME\n","\n","TO DO 9 : DANCE \n","```\n","\n","# 9 | Ending 🏁\n","\n","**THAT'S IT FOR TODAY GUYS**\n","\n","**WE WILL GO DEEPER INTO THE DATA IN THE UPCOMING VERSIONS**\n","\n","**PLEASE COMMENT YOUR THOUGHTS, HIHGLY APPRICIATED**\n","\n","**DONT FORGET TO MAKE AN UPVOTE, IF YOU LIKED MY WORK $:)$**\n","\n","<img src = \"https://i.imgflip.com/19aadg.jpg\">\n","\n","**PEACE OUT $:)$**"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.10"},"papermill":{"default_parameters":{},"duration":43.659317,"end_time":"2023-07-04T07:36:08.367326","environment_variables":{},"exception":null,"input_path":"__notebook__.ipynb","output_path":"__notebook__.ipynb","parameters":{},"start_time":"2023-07-04T07:35:24.708009","version":"2.4.0"}},"nbformat":4,"nbformat_minor":5}