{"cells":[{"source":"<a href=\"https://www.kaggle.com/code/ayushs9020/utc-pytorch-dataloader-cafa?scriptVersionId=135425683\" target=\"_blank\"><img align=\"left\" alt=\"Kaggle\" title=\"Open in Kaggle\" src=\"https://kaggle.com/static/images/open-in-kaggle.svg\"></a>","metadata":{},"cell_type":"markdown"},{"cell_type":"markdown","id":"bc6e80da","metadata":{"papermill":{"duration":0.006079,"end_time":"2023-07-01T12:54:48.258422","exception":false,"start_time":"2023-07-01T12:54:48.252343","status":"completed"},"tags":[]},"source":["# CAFA 5 Protein Prediction\n","\n","<img src = \"https://i.pinimg.com/originals/8e/de/73/8ede737acdcf671e000ae0f87a742e40.png\" width = 400>\n","\n","The `goal` of this competition is to `predict the function of a set of proteins`. We will `develop a model trained` on the `amino-acid sequences` of the `proteins and on other data`. Our work `will help researchers` better `understand the function of proteins`, which is `important for discovering` `how cells, tissues, and organs work`.\n","\n","<a herf = \"https://www.kaggle.com/code/gusthema/cafa-5-protein-function-with-tensorflow/notebook\"><img src = \"https://png.pngtree.com/png-clipart/20230401/original/pngtree-get-started-banner-png-image_9017831.png\" width = 200></a>\n","\n","# 1 | Basic Terminologies üíª\n","\n","* $Structure$ $of$ $a$ $Protein$\n","* $Gene$ $Ontology$ $(GO)$\n","\n","## 1.1 | Structure of a Protein\n","\n","So what is a actually a **Protein...?**\n","\n","First of all lets understand the structure of an **Atom**\n","\n","<img src = \"https://www.sciencefacts.net/wp-content/uploads/2020/11/Parts-of-an-Atom-Diagram.jpg\"  width = 300>\n","\n","There is a really good image I found of the `structure of atom`. Though there are many debates on the structure like this, but this `model is accepted universaly at this moment`.\n","\n","In the centre we have the `Neucleus`. The `Neucleus` is made up of $2$ more structures named as `Neutron` and `Proton`. A `Proton` is `positively charged element` and a `Neutron` is a `neutral charged element`. A `Electron`, `negatively charged element`, `orbits` this `Neucleus` at some `distance apart`.\n","\n","The `more we increase the number` of `Electrons` and `Protons`. The `bigger the atoms becomes`.\n","\n","There are `different shells` where the `Electrons reside`. The `more closer the shell` is, the `less Electrons` it contrains. There are mainly $4$ shells. \n","\n","|||\n","|---|---\n","|$K$|$2$\n","|$L$|$8$\n","|$M$|$18$\n","|$N$|$32$\n","\n","Once an atom `fills its outer most shell` with `Electrons`. It becomes `stable atom` and try to `refuse any donation` or `recieve of extra atom`.\n","\n","<img src = \"https://cdn1.byjus.com/wp-content/uploads/2022/01/word-image128.png\" width = 400>\n","\n","`Different atoms combine` to `share Electrons` and become `stable Molecules` \n","\n","<img src = \"https://www.astrochem.org/sci_img/Amino_Acid_Structure.jpg\" width = 300>\n","\n","A `Amino Acid` is made up of mainly $4$ different atoms\n","`[H , C , O , N]`. \n","\n","<img src = \"https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/L-amino_acid_structure.svg/1200px-L-amino_acid_structure.svg.png\" alt = \"Bro use Light Theme\" width = 300 >\n","\n","We also have a free `Electron Pair` of `Carbon` in this molecule, we call this as a `Side Chain` which can be of different types. Basically this `Side Chain` provide the flexibility to make `different types` of `Amino Acids`. This flexibilty allows for $20$ `different` `Amino Acids` \n","\n","When we join `Amino Acids` with `peptide bonds`, we get `Proteins`. Conncecting different types of `Amino Acids` ends up in different types of `Proteins`.\n","\n","## 1.2 | Gene Ontology (GO)\n","\n","$Gene$ $Ontology$ $(GO)$ is a `controlled vocabulary` that `describes the functions of genes` and gene products. It is constantly being updated as new information becomes available.\n","\n","There are mainly $3$ `Ontologies`\n","\n","||||\n","|---|---|---\n","|$Biological$ $Process$|Describes the `biological processes`|A gene product might be involved in the process of `cell cycle`/`signal transduction.`\n","|$Cellular$ $Component$|Describes the `cellular components`|A gene product might be located in the `nucleus`/`cytoplasm`.\n","|$Molecular$ $Function$|Describes the `molecular functions`|A gene product might be involved in the `catalysis of a reaction`/`binding of a molecule`.\n","\n","**[Gene Ontology Documentation](http://geneontology.org/docs/ontology-documentation/)**"]},{"cell_type":"markdown","id":"4e38eca2","metadata":{"papermill":{"duration":0.00548,"end_time":"2023-07-01T12:54:48.269475","exception":false,"start_time":"2023-07-01T12:54:48.263995","status":"completed"},"tags":[]},"source":["# 2 | Data üìä"]},{"cell_type":"code","execution_count":1,"id":"2743fa7c","metadata":{"_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.execute_input":"2023-07-01T12:54:48.282466Z","iopub.status.busy":"2023-07-01T12:54:48.28183Z","iopub.status.idle":"2023-07-01T12:54:48.365751Z","shell.execute_reply":"2023-07-01T12:54:48.364787Z"},"papermill":{"duration":0.093371,"end_time":"2023-07-01T12:54:48.368413","exception":false,"start_time":"2023-07-01T12:54:48.275042","status":"completed"},"tags":[]},"outputs":[],"source":["import pandas as pd \n","import re\n","from Bio.Seq import Seq"]},{"cell_type":"markdown","id":"16b493ae","metadata":{"papermill":{"duration":0.00516,"end_time":"2023-07-01T12:54:48.379109","exception":false,"start_time":"2023-07-01T12:54:48.373949","status":"completed"},"tags":[]},"source":["The $Training$ $Set$ contains all `proteins with annotated terms` that have been validated by \n","* $Experimental$\n","* $High-Throughput$ $Evidence$\n","* [$Traceable$ $Author$ $Statement$](https://wiki.geneontology.org/index.php/Traceable_Author_Statement_(TAS)#:~:text=The%20TAS%20evidence%20code%20covers,annotations%20come%20from%20review%20articles.)\n","* [$Inferred$ $by$ $Curator$ $(IC)$](https://wiki.geneontology.org/Inferred_by_Curator_(IC)) \n","\n","**Any other sources of Data are allowed**\n","\n","### 2.1.1.1 | Go-Basic.obo\n","\n","The $Ontology$ data is in the `file go-basic.obo`. This file is in $OBO$ `Biology-Oriented Language`. The nodes in `the graph are indexed` by the `term name`\n","```\n","subontology_roots = {'BPO':'GO:0008150',\n","                     'CCO':'GO:0005575',\n","                     'MFO':'GO:0003674'}\n","```"]},{"cell_type":"code","execution_count":2,"id":"2cad65a0","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-01T12:54:48.392088Z","iopub.status.busy":"2023-07-01T12:54:48.391324Z","iopub.status.idle":"2023-07-01T12:54:52.796284Z","shell.execute_reply":"2023-07-01T12:54:52.794916Z"},"jupyter":{"source_hidden":true},"papermill":{"duration":4.414452,"end_time":"2023-07-01T12:54:52.798948","exception":false,"start_time":"2023-07-01T12:54:48.384496","status":"completed"},"tags":[]},"outputs":[{"name":"stdout","output_type":"stream","text":["[Term]\n","id: GO:0000001\n","name: mitochondrion inheritance\n","namespace: biological_process\n","def: \"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\" [GOC:mcc, PMID:10873824, PMID:11389764]\n","synonym: \"mitochondrial inheritance\" EXACT []\n","is_a: GO:0048308 ! organelle inheritance\n","is_a: GO:0048311 ! mitochondrion distribution\n","\n"]}],"source":["with open('/kaggle/input/cafa-5-protein-function-prediction/Train/go-basic.obo') as file :\n","    \n","    content = file.read()\n","    stanzas =  re.findall(r'\\[Term\\][\\s\\S]*?(?=\\n\\[|$)' , content)\n","    \n","print(stanzas[0])"]},{"cell_type":"markdown","id":"42f3d0c4","metadata":{"papermill":{"duration":0.005367,"end_time":"2023-07-01T12:54:52.809978","exception":false,"start_time":"2023-07-01T12:54:52.804611","status":"completed"},"tags":[]},"source":["### 2.1.1.2 | Training Sequences.fasta\n","\n","This file contains only `sequences` for `proteins` with `annotations` in the dataset `labeled proteins`.\n","\n","This files are in `FASTA` format. \n","\n","This file contains . To obtain the full set of protein sequences for unlabeled proteins, the Swiss-Prot and TrEMBL databases can be found here."]},{"cell_type":"code","execution_count":3,"id":"8f7d510b","metadata":{"execution":{"iopub.execute_input":"2023-07-01T12:54:52.824041Z","iopub.status.busy":"2023-07-01T12:54:52.823654Z","iopub.status.idle":"2023-07-01T12:54:53.823055Z","shell.execute_reply":"2023-07-01T12:54:53.822016Z"},"papermill":{"duration":1.009905,"end_time":"2023-07-01T12:54:53.825389","exception":false,"start_time":"2023-07-01T12:54:52.815484","status":"completed"},"tags":[]},"outputs":[{"data":{"text/plain":["'>P20536 sp|P20536|UNG_VACCC Uracil-DNA glycosylase OS=Vaccinia virus (strain Copenhagen) OX=10249 GN=UNG PE=1 SV=1\\nMNSVTVSHAPYTITYHDDWEPVMSQLVEFYNEVASWLLRDETSPIPDKFFIQLKQPLRNK\\nRVCVCGIDPYPKDGTGVPFESPNF'"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":["with open(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_sequences.fasta\" , \"r\") as file:\n","    x = file.read()\n","    \n","x[:200]"]},{"cell_type":"markdown","id":"1ebe6691","metadata":{"papermill":{"duration":0.005442,"end_time":"2023-07-01T12:54:53.836661","exception":false,"start_time":"2023-07-01T12:54:53.831219","status":"completed"},"tags":[]},"source":["### 2.1.1.3 | Train Terms.tsv\n","\n","This file contains the list of annotated terms `ground truth` for the proteins in `train_sequences.fasta`. "]},{"cell_type":"code","execution_count":4,"id":"992b4191","metadata":{"execution":{"iopub.execute_input":"2023-07-01T12:54:53.850546Z","iopub.status.busy":"2023-07-01T12:54:53.849837Z","iopub.status.idle":"2023-07-01T12:54:57.762837Z","shell.execute_reply":"2023-07-01T12:54:57.761726Z"},"papermill":{"duration":3.922809,"end_time":"2023-07-01T12:54:57.765289","exception":false,"start_time":"2023-07-01T12:54:53.84248","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>term</th>\n","      <th>aspect</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0008152</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0034655</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0072523</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0044270</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>A0A009IHW8</td>\n","      <td>GO:0006753</td>\n","      <td>BPO</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>5363858</th>\n","      <td>X5L565</td>\n","      <td>GO:0050649</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363859</th>\n","      <td>X5L565</td>\n","      <td>GO:0016491</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363860</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005515</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363861</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0005488</td>\n","      <td>MFO</td>\n","    </tr>\n","    <tr>\n","      <th>5363862</th>\n","      <td>X5M5N0</td>\n","      <td>GO:0003674</td>\n","      <td>MFO</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5363863 rows √ó 3 columns</p>\n","</div>"],"text/plain":["            EntryID        term aspect\n","0        A0A009IHW8  GO:0008152    BPO\n","1        A0A009IHW8  GO:0034655    BPO\n","2        A0A009IHW8  GO:0072523    BPO\n","3        A0A009IHW8  GO:0044270    BPO\n","4        A0A009IHW8  GO:0006753    BPO\n","...             ...         ...    ...\n","5363858      X5L565  GO:0050649    MFO\n","5363859      X5L565  GO:0016491    MFO\n","5363860      X5M5N0  GO:0005515    MFO\n","5363861      X5M5N0  GO:0005488    MFO\n","5363862      X5M5N0  GO:0003674    MFO\n","\n","[5363863 rows x 3 columns]"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_terms.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"25ece2b3","metadata":{"papermill":{"duration":0.005899,"end_time":"2023-07-01T12:54:57.77728","exception":false,"start_time":"2023-07-01T12:54:57.771381","status":"completed"},"tags":[]},"source":["The first column indicates the `protein's UniProt accession ID`, the second is the `GO term ID`, and the third indicates in which `ontology the term appears`.\n","\n","### 2.1.1.4 | Train Taxonomy.tsv\n","\n","This file contains the list of `proteins and the species to which they belong`, represented by a `taxonomic identifier` `taxon ID` number."]},{"cell_type":"code","execution_count":5,"id":"ff1ca933","metadata":{"execution":{"iopub.execute_input":"2023-07-01T12:54:57.79095Z","iopub.status.busy":"2023-07-01T12:54:57.790565Z","iopub.status.idle":"2023-07-01T12:54:57.909771Z","shell.execute_reply":"2023-07-01T12:54:57.908707Z"},"papermill":{"duration":0.129109,"end_time":"2023-07-01T12:54:57.91234","exception":false,"start_time":"2023-07-01T12:54:57.783231","status":"completed"},"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>EntryID</th>\n","      <th>taxonomyID</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Q8IXT2</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Q04418</td>\n","      <td>559292</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>A8DYA3</td>\n","      <td>7227</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Q9UUI3</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q57ZS4</td>\n","      <td>185431</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>142241</th>\n","      <td>Q5TD07</td>\n","      <td>9606</td>\n","    </tr>\n","    <tr>\n","      <th>142242</th>\n","      <td>A8BB17</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142243</th>\n","      <td>A0A2R8QBB1</td>\n","      <td>7955</td>\n","    </tr>\n","    <tr>\n","      <th>142244</th>\n","      <td>P0CT72</td>\n","      <td>284812</td>\n","    </tr>\n","    <tr>\n","      <th>142245</th>\n","      <td>Q9NZ43</td>\n","      <td>9606</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>142246 rows √ó 2 columns</p>\n","</div>"],"text/plain":["           EntryID  taxonomyID\n","0           Q8IXT2        9606\n","1           Q04418      559292\n","2           A8DYA3        7227\n","3           Q9UUI3      284812\n","4           Q57ZS4      185431\n","...            ...         ...\n","142241      Q5TD07        9606\n","142242      A8BB17        7955\n","142243  A0A2R8QBB1        7955\n","142244      P0CT72      284812\n","142245      Q9NZ43        9606\n","\n","[142246 rows x 2 columns]"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":["pd.read_csv(\"/kaggle/input/cafa-5-protein-function-prediction/Train/train_taxonomy.tsv\" , sep = \"\\t\")"]},{"cell_type":"markdown","id":"642ab4a5","metadata":{"papermill":{"duration":0.006129,"end_time":"2023-07-01T12:54:57.924769","exception":false,"start_time":"2023-07-01T12:54:57.91864","status":"completed"},"tags":[]},"source":["### 2.1.1.5 | IA.txt\n","\n","IA.txt contains the information accretion (weights) for each GO term. These weights are used to compute weighted precision and recall, as described in the Evaluation section. \n","\n","## 2.1.2 | Test Set\n","\n","The $Test$ $Set$ is `unknown at the beginning` of the competition. It will contain `protein sequences` `their functions` from the `test superset` that `gained experimental annotations` between the `submission-deadline` and the `time of evaluation`.\n","\n","# 3 | PyTorch DataLoader ‚öôÔ∏è"]},{"cell_type":"code","execution_count":6,"id":"5bb14c0d","metadata":{"_kg_hide-input":true,"execution":{"iopub.execute_input":"2023-07-01T12:54:57.939303Z","iopub.status.busy":"2023-07-01T12:54:57.938882Z","iopub.status.idle":"2023-07-01T12:55:01.326457Z","shell.execute_reply":"2023-07-01T12:55:01.3254Z"},"papermill":{"duration":3.398152,"end_time":"2023-07-01T12:55:01.329232","exception":false,"start_time":"2023-07-01T12:54:57.93108","status":"completed"},"tags":[]},"outputs":[],"source":["import numpy as np\n","import tqdm\n","\n","from torch.utils.data import Dataset"]},{"cell_type":"markdown","id":"99c00b47","metadata":{"papermill":{"duration":0.006059,"end_time":"2023-07-01T12:55:01.341846","exception":false,"start_time":"2023-07-01T12:55:01.335787","status":"completed"},"tags":[]},"source":["The dataloader is highly inspired by **[Henri Upton](https://www.kaggle.com/henriupton)=>[ProteiNet üß¨ PyTorch+EMS2/T5/ProtBERT Embeddings](https://www.kaggle.com/code/henriupton/proteinet-pytorch-ems2-t5-protbert-embeddings)**"]},{"cell_type":"code","execution_count":7,"id":"dcfad9e0","metadata":{"execution":{"iopub.execute_input":"2023-07-01T12:55:01.356946Z","iopub.status.busy":"2023-07-01T12:55:01.355725Z","iopub.status.idle":"2023-07-01T12:55:01.368348Z","shell.execute_reply":"2023-07-01T12:55:01.367358Z"},"papermill":{"duration":0.02264,"end_time":"2023-07-01T12:55:01.37075","exception":false,"start_time":"2023-07-01T12:55:01.34811","status":"completed"},"tags":[]},"outputs":[],"source":["class P_Dataset(Dataset):\n","    \n","    def __init__(self , datatype ):\n","        super(P_Dataset).__init__()\n","        \n","        self.datatype = datatype\n","        \n","        embeds = np.load(\"/kaggle/input/t5embeds/test_embeds.npy\")\n","        ids = np.load(\"/kaggle/input/t5embeds/test_ids.npy\")\n","        \n","        embeds_list = [embeds[l,:] for l in range(embeds.shape[0])]\n","        \n","        self.df = pd.DataFrame(data = {\"EntryID\" : ids , \"embed\" : embeds_list})\n","        \n","        if datatype == \"train\":\n","            \n","            np_labels = np.load(\"/kaggle/input/train-targets-top500/train_targets_top500.npy\")[:len(embeds_list)]\n","            \n","            labels_df = self.df[\"EntryID\"]\n","            labels_df = pd.concat([labels_df , pd.DataFrame([row for row in np_labels])] , axis = 1)\n","#\n","            \n","            self.df = self.df.merge(pd.DataFrame(labels_df) , on=\"EntryID\")\n","            \n","    def __len__(self):return len(self.df)\n","    \n","    def __getitem__(self , index):\n","        \n","        embed = torch.tensor(self.df.iloc[index][\"embed\"] , dtype = torch.float32)\n","        \n","        if self.datatype == \"train\":\n","            \n","            targets = torch.tensor(self.df.iloc[index][\"labels_vect\"], dtype = torch.float32)\n","            \n","            return embed, targets\n","        \n","        id = self.df.iloc[index][\"EntryID\"]\n","        \n","        return embed, id"]},{"cell_type":"code","execution_count":8,"id":"2a924060","metadata":{"execution":{"iopub.execute_input":"2023-07-01T12:55:01.385287Z","iopub.status.busy":"2023-07-01T12:55:01.384881Z","iopub.status.idle":"2023-07-01T12:56:05.61034Z","shell.execute_reply":"2023-07-01T12:56:05.609294Z"},"papermill":{"duration":64.235901,"end_time":"2023-07-01T12:56:05.613141","exception":false,"start_time":"2023-07-01T12:55:01.37724","status":"completed"},"tags":[]},"outputs":[],"source":["train_data = P_Dataset(datatype = \"train\")"]},{"cell_type":"markdown","id":"e517a3db","metadata":{"papermill":{"duration":0.00609,"end_time":"2023-07-01T12:56:05.626292","exception":false,"start_time":"2023-07-01T12:56:05.620202","status":"completed"},"tags":[]},"source":["# 4 | TO DO LIST üìÑ\n","\n","```\n","TO DO 1 : VISUALIZE THE DATA\n","\n","TO DO 2 : TRAIN A MODEL\n","\n","TO DO 3 : TRY DIFFERENT MODELS\n","\n","TO DO 4 : ADD WANDB SUPPORT\n","\n","TO DO 5 : ADD TENSORFLOW DATA LOADER\n","\n","TO DO 6 : TRAIN A TF MODEL\n","\n","TO DO 7 : IMPROVE RESULTS\n","\n","TO DO 8 : DECREASE TRAINING TIME\n","\n","TO DO 9 : DANCE \n","```\n","\n","# 3 | Ending üèÅ\n","\n","**THAT'S IT FOR TODAY GUYS**\n","\n","**WE WILL GO DEEPER INTO THE DATA IN THE UPCOMING VERSIONS**\n","\n","**PLEASE COMMENT YOUR THOUGHTS, HIHGLY APPRICIATED**\n","\n","**DONT FORGET TO MAKE AN UPVOTE, IF YOU LIKED MY WORK $:)$**\n","\n","<img src = \"https://i.imgflip.com/19aadg.jpg\">\n","\n","**PEACE OUT $:)$**"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.10"},"papermill":{"default_parameters":{},"duration":90.339237,"end_time":"2023-07-01T12:56:07.158413","environment_variables":{},"exception":null,"input_path":"__notebook__.ipynb","output_path":"__notebook__.ipynb","parameters":{},"start_time":"2023-07-01T12:54:36.819176","version":"2.4.0"}},"nbformat":4,"nbformat_minor":5}